pkg/endpoint/bpf.go

```diff
@@ -228,88 +228,77 @@ func (p *proxyPolicy) GetListener() string {
-					// with different port values. The redirect will be created
-					// when the mapping is available or when the port name
-					// conflicts have been resolved in POD specs.
-					continue
-				}
-				// desiredRedirects starts out empty, so we can use it check
-				// if the redirect has already been updated on this round.
-				if desiredRedirects[proxyID] != 0 {
-					continue
-				}
-
-				var redirectPort uint16
-
-				pp := e.newProxyPolicy(l4, v, dstPort, dstProto)
-				proxyPort, err, finalizeFunc, revertFunc := e.proxy.CreateOrUpdateRedirect(e.aliveCtx, &pp, proxyID, e, proxyWaitGroup)
-				if err != nil {
-					// Skip redirects that can not be created or updated.  This
-					// can happen when a listener is missing, for example when
-					// restarting and k8s delivers the CNP before the related
-					// CEC.
-					// Policy is regenerated when listeners are added or removed
-					// to fix this condition when the listener is available.
-					e.getLogger().WithField(logfields.Listener, pp.GetListener()).WithError(err).Debug("Redirect rule with missing listener skipped, will be applied once the listener is available")
-					continue
-				}
-				redirectPort = proxyPort
-				finalizeList.Append(finalizeFunc)
-				revertStack.Push(revertFunc)
-				desiredRedirects[proxyID] = redirectPort
+	selectorPolicy.CreateRedirects(func(l4 *policy.L4Filter) error {
+		for _, v := range l4.PerSelectorPolicies {
+			if v == nil || !v.IsRedirect() {
+				continue
+			}
+			// proxyID() returns also the destination port for the policy,
+			// which may be resolved from a named port
+			proxyID, dstPort, dstProto := e.proxyID(l4, v.Listener)
+			if proxyID == "" {
+				// Skip redirects for which a proxyID cannot be created.
+				// This may happen due to the named port mapping not
+				// existing or multiple PODs defining the same port name
+				// with different port values. The redirect will be created
+				// when the mapping is available or when the port name
+				// conflicts have been resolved in POD specs.
+				continue
+			}
+			// desiredRedirects starts out empty, so we can use it check
+			// if the redirect has already been updated on this round.
+			if desiredRedirects[proxyID] != 0 {
+				continue
+			}
 
-				// Update the endpoint API model to report that Cilium manages a
-				// redirect for that port.
-				statsKey := policy.ProxyStatsKey(l4.Ingress, string(l4.Protocol), dstPort, redirectPort)
-				proxyStats := e.getProxyStatistics(statsKey, string(l4.L7Parser), dstPort, l4.Ingress, redirectPort)
+			var redirectPort uint16
 
-				updatedStats = append(updatedStats, proxyStats)
-			}
-			// Can skip policy map updates when a new policy has not been computed,
-			// as the update will be done via applyPolicyMapChanges.
-			// Returning nil will cause desired policy map update to be skipped by
-			// the caller.
-			if e.desiredPolicy == e.realizedPolicy {
-				return nil
+			pp := e.newProxyPolicy(l4, v, dstPort, dstProto)
```

Comment:
`e.newProxyPolicy()` does not reference `e` at all. Can we get rid of it? In fact, can we get rid of that type altogether, and move it to `pkg/proxy`?

Reply:
This type replaced passing `*policy.L4Filter` directly to the proxy package, and was implemented at the endpoint package due to named port resolution happening there. Having the endpoint as a receiver for `newProxyPolicy` is likely an historical artifact, removed. IMO moving the implementation to the proxy package would be odd as it uses this via `endpoint.EndpointUpdater`, expecting the implementation to be outside of the package.

Reply:
Reasonable enough. Thanks for the cleanup.